source/byte_buf.c

```diff
@@ -162,48 +163,44 @@ bool aws_byte_cursor_next_split(
     char split_on,
     struct aws_byte_cursor *AWS_RESTRICT substr) {
 
-    bool first_run = false;
-    if (!substr->ptr) {
-        first_run = true;
-        substr->ptr = input_str->ptr;
-        substr->len = 0;
-    }
-
-    if (substr->ptr > input_str->ptr + input_str->len) {
-        /* This will hit if the last substring returned was an empty string after terminating split_on. */
-        AWS_ZERO_STRUCT(*substr);
-        return false;
-    }
-
-    /* Calculate first byte to search. */
-    substr->ptr += substr->len;
-    /* Remaining bytes is the number we started with minus the number of bytes already read. */
-    substr->len = input_str->len - (substr->ptr - input_str->ptr);
-
-    if (!first_run && substr->len == 0) {
-        /* This will hit if the string doesn't end with split_on but we're done. */
-        AWS_ZERO_STRUCT(*substr);
-        return false;
-    }
-
-    if (!first_run && *substr->ptr == split_on) {
-        /* If not first rodeo and the character after substr is split_on, skip. */
-        ++substr->ptr;
-        --substr->len;
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(input_str));
+
+    /* If substr is zeroed-out, then this is the first run. */
+    const bool first_run = substr->ptr == NULL;
+    if (first_run) {
+        if (input_str->ptr) {
+            *substr = *input_str;
+        } else {
+            /* NOTE: It's also legal for input_str to be zeroed out.
+             * In that case, set substr->ptr to something else so the next split() call doesn't look like first run. */
+            substr->ptr = (void *)"";
+            substr->len = 0;
+        }
 
-        if (substr->len == 0) {
-            /* If split character was last in the string, return empty substr. */
-            return true;
+    } else {
+        /* This is not the first run.
+         * Advance substr past the previous split. */
+        const uint8_t *input_end = input_str->ptr + input_str->len;
+        substr->ptr += substr->len + 1;
+        if (substr->ptr > input_end || substr->ptr < input_str->ptr) { /* 2nd check is overflow check */
```

Comment:
`substr->ptr >= input_end` ? Note that input_end already points to 1 after the last char of input_str

Reply:
Yeah, this is on purpose. If the `split_on` char is the last char in the string, then we want the last split to be "", where the ptr is actually pointing at `input_end`.

For example, given "12&34&", split on '&', we should get the following 3 splits: "12", "34", ""
That final "" is pointing at `input_end` (but with len 0 so the user never actually accesses the memory)

Reply:
Here is my attempt to explain, in carefully aligned ascii art:
```
address:  1234567
input:   "AB&CD&"  ptr:1 len:6 input_end:7
split0:  "AB"      ptr:1 len:2
split1:     "CD"   ptr:4 len:2
split2:        ""  ptr:7 len:0
```